### PR TITLE
Jira Link Workflow Regex Improvement

### PR DESCRIPTION
**Pull request by MichaelYochpaz**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
relates: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-5963)

## Description
This PR improves the regex for matching Jira URLs on GitHub PRs.

The new regex:
1. Is case-insensitive (both for `fixes` / `relates`, and for the URL itself).
2. Captures `http` scheme URLs (when entering our Jira instance with `http` it redirects to `https`, so while an `http` URL usage is unlikely in a PR, it's still valid).
3. Has escaping for the `.` character, which was missing from the URL.

Usage of `re.findall` which returns strings or a tuple of strings if there are match groups, is replaced with `re.finditer`, which returns `Match` objects and allows much more control and includes additional metadata for the found items.

Plus a few minor improvements to the code (mostly aesthetic, the behavior is mostly unchanged).

[Previous Regex (regex101)](https://regex101.com/r/8FWrOE/1)
[New Regex (regex101)](https://regex101.com/r/0Huq7z/1)

### CODE REVIEW

File: Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py [hunk 1]
```diff
@@ -40,21 +40,21 @@ def find_fixed_issue_in_body(body_text, is_merged):
...
-    fixed_issue = [{"link": link, "id": issue_id} for link, issue_id in fixed_jira_issues]
-    related_issue = []
+    issues = [
+        {"link": x.group(1), "id": x.group(2)} for x in re.finditer(JIRA_FIXED_ISSUE_REGEX, body_text)
+    ]
 
     if not is_merged:
-        print("Not merging, getting related issues.")
-        related_issue = [{"link": link, "id": issue_id} for link, issue_id in related_jira_issue]
+        print(f"Not merging, collecting related issues.")
```
Comment (bziser) [inline Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py:51]:
```suggestion
        print("Not merging, collecting related issues.")
```

File: Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py [hunk 2]
```diff
@@ -70,10 +70,7 @@ def trigger_generic_webhook(options):
     print(f"Detected Pr: {pr_title=}, {pr_link=}, {pr_body=}")
 
     # Handle cases where the PR did not intend to add links:
-    if ("fixes:" not in pr_body.lower()
-            and "relates:" not in pr_body.lower()
-            and "fixed:" not in pr_body.lower()
-            and "related:" not in pr_body.lower()):
+    if not any(x.casefold() in pr_body.casefold() for x in ("fixes:", "relates:", "fixed:", "related:")):
```
Comment (bziser) [inline Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py:73]:
```suggestion
        if all(x.casefold() not in pr_body.casefold() for x in ("fixes:", "relates:", "fixed:", "related:")):
```

File: Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py [hunk 3]
```diff
@@ -40,21 +40,21 @@ def find_fixed_issue_in_body(body_text, is_merged):
     Getting the issues url in the PR's body as part of `fixing: <issue>` format.
     Return list of issues found: [{"link": link, "id": issue_id}]
     """
-    fixed_jira_issues = re.findall(JIRA_FIXED_ISSUE_REGEX, body_text)
-    related_jira_issue = re.findall(JIRA_RELATED_ISSUE_REGEX, body_text)
-    print(f'Detected {related_jira_issue=}')
-
     # If a PR is not merged, we just add the pr link to all the linked issues using Gold.
     # If the PR is merged, we only send issues that should be closed by it.
     # Assuming If the PR was merged, all the related links were fetched when the PR last edited.
```
Comment (bziser) [inline Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py:45]: Maybe better add it to the docstring ?

**Review by bziser — COMMENTED** [review]
👏

File: Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py (hunk 3 above)
Comment (MichaelYochpaz) [reply Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py:45]: Good idea

File: Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py [hunk 4]
```diff
@@ -70,10 +74,7 @@ def trigger_generic_webhook(options):
     print(f"Detected Pr: {pr_title=}, {pr_link=}, {pr_body=}")
 
     # Handle cases where the PR did not intend to add links:
-    if ("fixes:" not in pr_body.lower()
-            and "relates:" not in pr_body.lower()
-            and "fixed:" not in pr_body.lower()
-            and "related:" not in pr_body.lower()):
+    if all(x.casefold() not in pr_body.casefold() for x in ("fixes:", "relates:", "fixed:", "related:")):
```
Comment (MichaelYochpaz) [inline Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py:77]: `fixed:` and `related:` are not in the regex, so are these cases even valid? (need to check)

Comment (bziser) [reply Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py:77]: It is important not to break backward, and until now, all 4 options have been viable.

**Comment by bziser** [comment]
Adding @JasBeilin to review it as she implemented this

**Comment by MichaelYochpaz** [comment]
Keeping it as draft atm as I want to test it before merging with a real PR and Jira ticket.

File: Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py [hunk 5]
```diff
@@ -7,10 +7,10 @@
 
 urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
 
-JIRA_HOST_FOR_REGEX = "https:\/\/jira-hq.paloaltonetworks.local\/browse\/"
-JIRA_KEY_REGEX = "([A-Z][A-Z0-9]+-[0-9]+))\s?"
-JIRA_FIXED_ISSUE_REGEX = f"[fF]ixes:\s?.*({JIRA_HOST_FOR_REGEX}{JIRA_KEY_REGEX}"
-JIRA_RELATED_ISSUE_REGEX = f"[rR]elates:\s?.*({JIRA_HOST_FOR_REGEX}{JIRA_KEY_REGEX}"
+JIRA_URL_REGEX = \
+    r"(?:\[.+\]\()?(?P<url>https?:\/\/jira-hq\.paloaltonetworks\.local\/browse\/(?P<issue_id>[a-zA-Z][a-zA-Z0-9]+-[0-9]+))"
```
Comment (JasBeilin) [inline Utils/github_workflow_scripts/jira_integration_scripts/link_pr_to_jira_issue.py:11]: just for the sake of keeping order- key and ID are different things

```suggestion
    r"(?:\[.+\]\()?(?P<url>https?:\/\/jira-hq\.paloaltonetworks\.local\/browse\/(?P<issue_key>[a-zA-Z][a-zA-Z0-9]+-[0-9]+))"
```

**Comment by bziser** [comment]
> Keeping it as draft atm as I want to test it before merging with a real PR and Jira ticket.

You can open a jira ticket for this issue (if there is no like this) and add in this pr to test like this :)

**Comment by MichaelYochpaz** [comment]
TODO: Assure using multiple "relates" / "fixes" ticket links works and connects them all to the PR.

**Pull request closed without merging**